proposals/key-equals.md

```diff
@@ -0,0 +1,133 @@
+
+# Ad hoc structural equality
+
+One of the component features proposed as a part of records is structural equality defined on certain members. This seems like a feature which could be orthogonal and useful for regular classes. The following proposal details how we could mark specific members as "keys" to a generated structural equality.
+
+Much like the `Key` modifier on anonymous properties in VB, a `key` modifier will be permitted on C# fields or properties.
+
+If there are any `key` modifiers present on any members in a class or struct declaration,
+equality will be generated for the type based on the annotated members as follows.
+
+If no `key` members are present, the type uses inherited equality, like today.
+
+## Synthesized members
+
+An example of a complex hierarchy with multiple types with `key` modifiers:
+
+```C#
+class A
+{
+    public key int P1 { get; }
+    public int P2 { get; }
+}
+class B : A
+{
+    public key int P3 { get; }
+    public int P4 { get; }
+}
+```
+
+produces
+
+```C#
+class A : IEquatable<A>
+{
+    public key int P1 { get; }
+    public int P2 { get; }
+
+    public override bool Equals(object o)
+        => o is A a && Equals(a);
+
+    public override int GetHashCode()
+        => P1;
+
+    public bool Equals(A a)
+        => EqualityContractOrigin == a.EqualityContractOrigin
+        && KeyEquals(a);
+
+    protected virtual Type EqualityContractOrigin
+        => typeof(A);
+
+    protected bool KeyEquals(A a)
+        => P1 == a.P1;
+
+    public static bool operator==(A left, A right) => left.Equals(right);
```

Comment:
This doesn't check for null?

Reply:
Good point! Yeah, it seems good to make this null-aware.